File: src/904. Fruit Into Baskets.js
/**
 * @param {number[]} fruits
 * @return {number}
 */

export function totalFruit(fruits) {
  const counts = new Map()
  let max = 0,
    left = 0
  for (let right = 0; right < fruits.length; right++) {
    counts.set(fruits[right], (counts.get(fruits[right]) || 0) + 1)
    while (counts.size > 2) {
      counts.set(fruits[left], counts.get(fruits[left]) - 1)
      if (counts.get(fruits[left]) === 0) counts.delete(fruits[left])
      left++
    }
    max = Math.max(max, right - left + 1)
  }
  return max
}

File: src/904. Fruit Into Baskets.cpp
class Solution {
public:
    int totalFruit(vector<int>& fruits) {
        int left, right;
        unordered_map<int, int> countMap;

        for (left = 0, right = 0; right < fruits.size(); ++right) {
            countMap[fruits[right]]++;
            if (countMap.size() <= 2) {
                continue;
            }
            countMap[fruits[left]]--;
            if (countMap[fruits[left]] == 0) {
                countMap.erase(fruits[left]);
            }
            left++;
        }

        return right - left;
    }
};